File: services/control-plane/src/storage-sqlite.py
import json
import os
import sqlite3
import sys
from pathlib import PurePosixPath


def fail(message: str, exit_code: int = 1) -> None:
    sys.stderr.write(message)
    sys.exit(exit_code)


def normalize_relative_path(target_path: str) -> str:
    normalized = PurePosixPath(target_path).as_posix().strip("/")
    if normalized in ("", "."):
        return ""
    if normalized.startswith("../") or normalized == "..":
        raise ValueError(f'Invalid storage path "{target_path}"')
    return normalized


def dir_key(relative_path: str) -> str:
    normalized = normalize_relative_path(relative_path)
    if normalized == "":
        return ""
    parent = PurePosixPath(normalized).parent.as_posix()
    return "" if parent == "." else parent


def basename_key(relative_path: str) -> str:
    normalized = normalize_relative_path(relative_path)
    if normalized == "":
        return ""
    return PurePosixPath(normalized).name


def row_path(parent_path: str, name: str) -> str:
    if not parent_path:
        return name
    return f"{parent_path}/{name}"


def ensure_schema(connection: sqlite3.Connection) -> None:
    connection.execute(
        """
        CREATE TABLE IF NOT EXISTS json_records (
          path TEXT PRIMARY KEY,
          parent_path TEXT NOT NULL,
          name TEXT NOT NULL,
          json_text TEXT NOT NULL
        )
        """
    )
    connection.execute(
        """
        CREATE TABLE IF NOT EXISTS logical_dirs (
          path TEXT PRIMARY KEY,
          parent_path TEXT NOT NULL,
          name TEXT NOT NULL
        )
        """
    )
    connection.execute(
        """
        INSERT OR IGNORE INTO logical_dirs(path, parent_path, name)
        VALUES('', '', '')
        """
    )
    connection.commit()


def ensure_dir(connection: sqlite3.Connection, relative_dir: str) -> None:
    normalized = normalize_relative_path(relative_dir)
    current = ""
    if normalized:
        for part in normalized.split("/"):
            next_path = row_path(current, part)
            connection.execute(
                """
                INSERT OR IGNORE INTO logical_dirs(path, parent_path, name)
                VALUES(?, ?, ?)
                """,
                (next_path, current, part),
            )
            current = next_path
    connection.commit()


def exists(connection: sqlite3.Connection, relative_path: str) -> bool:
    normalized = normalize_relative_path(relative_path)
    row = connection.execute(
        "SELECT 1 FROM json_records WHERE path = ?",
        (normalized,),
    ).fetchone()
    return row is not None


def list_dirs(connection: sqlite3.Connection, relative_dir: str) -> list[str]:
    normalized = normalize_relative_path(relative_dir)
    rows = connection.execute(
        """
        SELECT path
        FROM logical_dirs
        WHERE parent_path = ? AND path <> ?
        ORDER BY path
        """,
        (normalized, normalized),
    ).fetchall()
    return [row[0] for row in rows]


def list_json_files(connection: sqlite3.Connection, relative_dir: str) -> list[str]:
    normalized = normalize_relative_path(relative_dir)
    rows = connection.execute(
        """
        SELECT path
        FROM json_records
        WHERE parent_path = ?
        ORDER BY path
        """,
        (normalized,),
    ).fetchall()
    return [row[0] for row in rows]


def read_json(connection: sqlite3.Connection, relative_path: str):
    normalized = normalize_relative_path(relative_path)
    row = connection.execute(
        "SELECT json_text FROM json_records WHERE path = ?",
        (normalized,),
    ).fetchone()
    if row is None:
        raise FileNotFoundError(normalized)
    return json.loads(row[0])


def write_json(connection: sqlite3.Connection, relative_path: str, payload) -> None:
    normalized = normalize_relative_path(relative_path)
    parent = dir_key(normalized)
    ensure_dir(connection, parent)
    connection.execute(
        """
        INSERT INTO json_records(path, parent_path, name, json_text)
        VALUES(?, ?, ?, ?)
        ON CONFLICT(path) DO UPDATE SET
          parent_path = excluded.parent_path,
          name = excluded.name,
          json_text = excluded.json_text
        """,
        (
            normalized,
            parent,
            basename_key(normalized),
            json.dumps(payload, ensure_ascii=True, indent=2) + "\n",
        ),
    )
    connection.commit()


def remove_json(connection: sqlite3.Connection, relative_path: str) -> None:
    normalized = normalize_relative_path(relative_path)
    connection.execute("DELETE FROM json_records WHERE path = ?", (normalized,))
    connection.commit()


def absolute_result_paths(paths: list[str], data_dir: str) -> list[str]:
    return [os.path.normpath(os.path.join(data_dir, item)) for item in paths]


def main() -> None:
    raw = sys.stdin.read()
    if not raw.strip():
        fail("Missing sqlite storage request payload.")

    try:
        request = json.loads(raw)
    except json.JSONDecodeError as exc:
        fail(f"Invalid sqlite storage request JSON: {exc}")

    db_path = request.get("db_path")
    data_dir = request.get("data_dir")
    target_path = request.get("target_path")
    action = request.get("action")
    payload = request.get("payload")

    if not isinstance(db_path, str) or not db_path:
        fail("sqlite storage request missing db_path")
    if not isinstance(data_dir, str) or not data_dir:
        fail("sqlite storage request missing data_dir")
    if not isinstance(target_path, str):
        fail("sqlite storage request missing target_path")
    if not isinstance(action, str):
        fail("sqlite storage request missing action")

    os.makedirs(os.path.dirname(db_path), exist_ok=True)

    connection = sqlite3.connect(db_path)
    try:
        ensure_schema(connection)
        normalized = normalize_relative_path(target_path)

        if action == "ensure_dir":
            ensure_dir(connection, normalized)
            response = {"ok": True}
        elif action == "exists":
            response = {"ok": True, "exists": exists(connection, normalized)}
        elif action == "list_dirs":
            ensure_dir(connection, normalized)
            response = {
                "ok": True,
                "paths": absolute_result_paths(list_dirs(connection, normalized), data_dir),
            }
        elif action == "list_json_files":
            ensure_dir(connection, normalized)
            response = {
                "ok": True,
                "paths": absolute_result_paths(list_json_files(connection, normalized), data_dir),
            }
        elif action == "read_json":
            response = {"ok": True, "data": read_json(connection, normalized)}
        elif action == "write_json":
            write_json(connection, normalized, payload)
            response = {"ok": True}
        elif action == "remove_json":
            remove_json(connection, normalized)
            response = {"ok": True}
        else:
            response = {"ok": False, "error": f"Unsupported sqlite storage action: {action}"}

        sys.stdout.write(json.dumps(response))
    except Exception as exc:  # pragma: no cover - surfaced to caller
        sys.stdout.write(json.dumps({"ok": False, "error": str(exc)}))
        sys.exit(1)
    finally:
        connection.close()


if __name__ == "__main__":
    main()

File: services/control-plane/src/storage-sqlite.mjs
import fs from "node:fs";
import path from "node:path";
import { DatabaseSync } from "node:sqlite";

function normalizeRelativePath(value) {
  const normalized = String(value || "").replaceAll("\\", "/").replace(/^\/+|\/+$/g, "");
  if (!normalized || normalized === ".") return "";
  if (normalized === ".." || normalized.startsWith("../") || normalized.includes("/../")) {
    throw new Error(`Invalid storage path "${value}"`);
  }
  return normalized;
}

function parentPath(value) {
  const normalized = normalizeRelativePath(value);
  const index = normalized.lastIndexOf("/");
  return index < 0 ? "" : normalized.slice(0, index);
}

function baseName(value) {
  const normalized = normalizeRelativePath(value);
  const index = normalized.lastIndexOf("/");
  return index < 0 ? normalized : normalized.slice(index + 1);
}

function ensureSchema(database) {
  database.exec(`
    PRAGMA journal_mode = WAL;
    PRAGMA synchronous = FULL;
    CREATE TABLE IF NOT EXISTS json_records (
      path TEXT PRIMARY KEY,
      parent_path TEXT NOT NULL,
      name TEXT NOT NULL,
      json_text TEXT NOT NULL
    );
    CREATE TABLE IF NOT EXISTS logical_dirs (
      path TEXT PRIMARY KEY,
      parent_path TEXT NOT NULL,
      name TEXT NOT NULL
    );
    INSERT OR IGNORE INTO logical_dirs(path, parent_path, name) VALUES('', '', '');
  `);
}

function ensureDir(database, value) {
  const normalized = normalizeRelativePath(value);
  const insert = database.prepare(
    "INSERT OR IGNORE INTO logical_dirs(path, parent_path, name) VALUES(?, ?, ?)",
  );
  let current = "";
  for (const part of normalized ? normalized.split("/") : []) {
    const next = current ? `${current}/${part}` : part;
    insert.run(next, current, part);
    current = next;
  }
}

function absolutePaths(dataDir, rows) {
  return rows.map((row) => path.normalize(path.join(dataDir, row.path)));
}

function execute(database, request) {
  const target = normalizeRelativePath(request.target_path);
  switch (request.action) {
    case "apply_batch":
      database.exec("BEGIN IMMEDIATE");
      try {
        for (const operation of request.operations || []) {
          execute(database, {
            ...request,
            action: operation.kind === "write" ? "write_json" : "remove_json",
            target_path: operation.target_path,
            payload: operation.payload,
            operations: undefined,
          });
        }
        database.exec("COMMIT");
        return { ok: true };
      } catch (error) {
        database.exec("ROLLBACK");
        throw error;
      }
    case "ensure_dir":
      ensureDir(database, target);
      return { ok: true };
    case "exists":
      return {
        ok: true,
        exists: Boolean(database.prepare("SELECT 1 FROM json_records WHERE path = ?").get(target)),
      };
    case "list_dirs":
      ensureDir(database, target);
      return {
        ok: true,
        paths: absolutePaths(
          request.data_dir,
          database.prepare(
            "SELECT path FROM logical_dirs WHERE parent_path = ? AND path <> ? ORDER BY path",
          ).all(target, target),
        ),
      };
    case "list_json_files":
      ensureDir(database, target);
      return {
        ok: true,
        paths: absolutePaths(
          request.data_dir,
          database.prepare("SELECT path FROM json_records WHERE parent_path = ? ORDER BY path").all(target),
        ),
      };
    case "read_json": {
      const row = database.prepare("SELECT json_text FROM json_records WHERE path = ?").get(target);
      if (!row) throw new Error(`Storage record not found: ${target}`);
      return { ok: true, data: JSON.parse(row.json_text) };
    }
    case "write_json":
      ensureDir(database, parentPath(target));
      database.prepare(`
        INSERT INTO json_records(path, parent_path, name, json_text)
        VALUES(?, ?, ?, ?)
        ON CONFLICT(path) DO UPDATE SET
          parent_path = excluded.parent_path,
          name = excluded.name,
          json_text = excluded.json_text
      `).run(target, parentPath(target), baseName(target), `${JSON.stringify(request.payload, null, 2)}\n`);
      return { ok: true };
    case "remove_json":
      database.prepare("DELETE FROM json_records WHERE path = ?").run(target);
      return { ok: true };
    default:
      throw new Error(`Unsupported sqlite storage action: ${request.action}`);
  }
}

let database;
try {
  const request = JSON.parse(fs.readFileSync(0, "utf8"));
  fs.mkdirSync(path.dirname(request.db_path), { recursive: true });
  database = new DatabaseSync(request.db_path);
  ensureSchema(database);
  process.stdout.write(JSON.stringify(execute(database, request)));
} catch (error) {
  process.stdout.write(JSON.stringify({
    ok: false,
    error: error instanceof Error ? error.message : String(error),
  }));
  process.exitCode = 1;
} finally {
  database?.close();
}
